test/e2e/dra/dra.go

```diff
@@ -305,11 +559,391 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
+					pod1, err := f.ClientSet.CoreV1().Pods(pod1.Namespace).Get(ctx, pod1.Name, metav1.GetOptions{})
+					framework.ExpectNoError(err, "get first pod")
+					if pod1.Spec.NodeName == "" {
+						framework.Fail("first pod should be running on node, was not scheduled")
+					}
+					gomega.Expect(pod1.Spec.NodeName).ToNot(gomega.Equal(node), "first pod should run on different node than second one")
+					gomega.Expect(driver.Controller.GetNumDeallocations()).To(gomega.Equal(int64(1)), "number of deallocations")
+				})
+			})
+
+		}
+		ginkgo.Context("with node-local resources", func() {
+			driver := NewDriver(f, nodes, perNode(1, nodes))
+			driver.parameterMode = parameterMode
+			b := newBuilder(f, driver)
+
+			tests := func(allocationMode resourcev1alpha2.AllocationMode) {
+				ginkgo.It("uses all resources", func(ctx context.Context) {
+					objs, _ := b.flexibleParameters()
+					var pods []*v1.Pod
+					for i := 0; i < len(nodes.NodeNames); i++ {
+						pod, template := b.podInline(allocationMode)
+						pods = append(pods, pod)
+						objs = append(objs, pod, template)
+					}
+					b.create(ctx, objs...)
+
+					for _, pod := range pods {
+						err := e2epod.WaitForPodRunningInNamespace(ctx, f.ClientSet, pod)
+						framework.ExpectNoError(err, "start pod")
+					}
+
+					// The pods all should run on different
+					// nodes because the maximum number of
+					// claims per node was limited to 1 for
+					// this test.
+					//
+					// We cannot know for sure why the pods
+					// ran on two different nodes (could
+					// also be a coincidence) but if they
+					// don't cover all nodes, then we have
+					// a problem.
+					used := make(map[string]*v1.Pod)
+					for _, pod := range pods {
+						pod, err := f.ClientSet.CoreV1().Pods(pod.Namespace).Get(ctx, pod.Name, metav1.GetOptions{})
+						framework.ExpectNoError(err, "get pod")
+						nodeName := pod.Spec.NodeName
+						if other, ok := used[nodeName]; ok {
+							framework.Failf("Pod %s got started on the same node %s as pod %s although claim allocation should have been limited to one claim per node.", pod.Name, nodeName, other.Name)
+						}
+						used[nodeName] = pod
+					}
+				})
+			}
+
+			ginkgo.Context("with delayed allocation", func() {
+				tests(resourcev1alpha2.AllocationModeWaitForFirstConsumer)
+			})
+
+			ginkgo.Context("with immediate allocation", func() {
+				tests(resourcev1alpha2.AllocationModeImmediate)
+			})
+		})
+	}
+
+	tests := func(parameterMode parameterMode) {
+		ginkgo.Context("on single node", func() {
+			singleNodeTests(parameterMode)
+		})
+		ginkgo.Context("on multiple nodes", func() {
+			multiNodeTests(parameterMode)
+		})
+	}
+
+	ginkgo.Context("with ConfigMap parameters", func() { tests(parameterModeConfigMap) })
+	ginkgo.Context("with translated parameters", func() { tests(parameterModeTranslated) })
+	ginkgo.Context("with structured parameters", func() { tests(parameterModeStructured) })
```

Comment:
It looks like only one of these modes accepts the multiNodeTests. Better call singleNodeTests and multiNodeTests as necessary, instead of hiding the logic inside "tests".

Reply:
`tests` calls both and doesn't care what they run.

`multiNodeTests` contains "with node-local resources" and that shares the node setup with the other contexts that are under the `if parameterMode == parameterModeConfigMap` check.

I could invert the nesting, but I think that'll lead to more duplication.